### PR TITLE
Add new streamer gemm

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we add:
* adapt the streamer-gemm with the new integration mechanism
* add meaningful spatial strides in software test

### CODE REVIEW

File: target/snitch_cluster/sw/snax/streamer-gemm/src/snax-streamer-gemm-lib.c [hunk 1]
```diff
@@ -60,19 +61,32 @@ void set_streamer_start() { write_csr(981, 1); }
...
-    write_csr(986, 0);
+}
+
+// Read perforamcne counter of the Streamer, a read-only CSR
+uint32_t read_gemm_streamer_perf_counter() {
+    uint32_t perf_counter = read_csr(982);
+    return perf_counter;
+}
+
+// Read perforamcne counter of GEMM, a read-only CSR
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/streamer-gemm/src/snax-streamer-gemm-lib.c:88]:
```suggestion
// Read performance counter of GEMM, a read-only CSR
```

File: target/snitch_cluster/sw/snax/streamer-gemm/src/snax-streamer-gemm-lib.c [hunk 2]
```diff
@@ -60,19 +61,32 @@ void set_streamer_start() { write_csr(981, 1); }
...
 
 // Poll until Streamer and GEMM accelerator finish
 void wait_streamer_gemm() {
+    write_csr(987, 0);
+    write_csr(987, 0);
     write_csr(981, 0);
-    write_csr(986, 0);
+}
+
+// Read perforamcne counter of the Streamer, a read-only CSR
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/streamer-gemm/src/snax-streamer-gemm-lib.c:82]:
```suggestion
// Read performance counter of the Streamer, a read-only CSR
```

File: target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c [hunk 3]
```diff
@@ -36,13 +34,13 @@ int main() {
...
 
         // Set Streamer configuration CSR
-        set_streamer_csr(K, N, M, strideInnermostA, ldA, strideInnermostB, ldB,
-                         strideInnermostC, ldC, delta_local_a, delta_local_b,
-                         delta_local_c);
+        set_streamer_csr(K, N, M, strideInnermostA, ldA, spatialA,
+                         strideInnermostB, ldB, spatialB, strideInnermostC, ldC,
+                         spatialC, delta_local_a, delta_local_b, delta_local_c);
         // Set CSR to start Streamer
         set_streamer_start();
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c:45]: Dumb question: We can already setup the streamer, while the block gemm is not yet consuming the outputs of the streamer?
I would think it makes more sense to set all CSR settings first, and only then launch everything. But I'm assuming this is more optimal?

File: target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c [hunk 4]
```diff
@@ -36,13 +34,13 @@ int main() {
     // Wait for DMA to finish
     snrt_cluster_hw_barrier();
 
-    if (snrt_is_compute_core()) {
-        uint32_t gemm_start = snrt_mcycle();
+    if (snrt_global_core_idx() == 0) {
+        // uint32_t gemm_start = snrt_mcycle();
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c:38]: Why does this need a change?

**Review by JosseVanDelm — APPROVED** [review]
Hi, this looks great!

File: target/snitch_cluster/Makefile [hunk 5]
```diff
@@ -49,12 +49,15 @@ ifeq (${CFG_OVERRIDE}, cfg/snax-gemm.hjson)
 endif
 
 ifeq (${CFG_OVERRIDE}, cfg/snax-streamer-gemm.hjson)
-	SNAX_DEV_ROOT ?= $(shell $(BENDER) path snax-streamer-gemm-dev)
-	include $(SNAX_DEV_ROOT)/Makefile
+	SNAX_GEMM_ROOT ?= $(shell $(BENDER) path snax-gemm)
```
Comment (jorendumoulin) [inline target/snitch_cluster/Makefile:52]:
```suggestion
	SNAX_GEMM_ROOT ?= $(shell $(BENDER) path snax-streamer-gemm)
```
Is it possible to rename the bender target to `snax-streamer-gemm`? This way the name for the accelerator is fixed everywhere and will not cause confusion. Also then this Makefile script is easier to automate like in #112 .

**Review by jorendumoulin — COMMENTED** [review]
Looks good!

File: target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c (hunk 4 above)
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c:38]: Later we will add more compute cores and each of them will have an idx. Here we specify the core id to be consistent with multi-compute core situation.

File: target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c (hunk 3 above)
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c:45]: >Dumb question: We can already setup the streamer, while the block gemm is not yet consuming the outputs of the streamer?

Yes.

>I would think it makes more sense to set all CSR settings first, and only then launch everything. But I'm assuming this is more optimal?

We can start the streamer earlier so the streamer can already do some prefetch.

Comment (JosseVanDelm) [reply target/snitch_cluster/sw/apps/snax-streamer-gemm/src/snax-streamer-gemm.c:45]: Cool thanks!

**Comment by rgantonio** [comment]
Push this @xiaoling-yi 😄

**Pull request merged**